CRM/Contribute/Page/UserDashboard.php

```diff
@@ -142,6 +142,8 @@ public function run() {
     $invoiceSettings = Civi::settings()->get('contribution_invoice_settings');
     $invoicing = CRM_Utils_Array::value('invoicing', $invoiceSettings);
     $defaultInvoicePage = CRM_Utils_Array::value('default_invoice_page', $invoiceSettings);
+    $cs = CRM_Contact_BAO_Contact_Utils::generateChecksum($this->_contactId);
+    $this->assign('contactChecksum', $cs);
```

Comment:
Maybe squash these 2 lines into one?

Reply:
Also - does that regenerate it - I think if the user is here it is because one of the following is true
1) they are an admin user
2) they are the user who is being viewed & are logged in
3) they are the user who is being viewed & they got here using a checksum - which should be in the url ( TBH I thought once a check sum had been used it was stored to the session)

Adding the checksum to the url in scenario 1 seems risky -in 3 then it would be in the url & can be passed through?

Reply:
Not sure of I'm right but I think the change in Main.php already solves the problem of anonymous (constructed) PayNow links without a contact id in the url (payment links to be mailed).
The links in the dashboard itself probable do not need the checksum, but I thought it would be clearer to have one uniform url.
BTW It is not possible to reach the Contact Dashboard via a checksum link (since longtime on my wishlist)
So removing the checksum (if possibly required for security reasons you mention?) at least it would be good to leave the Contact id, because without it anonymous use of the link will fail like described in this issue.

Reply:
Actually maybe the anonymous link (to be mailed) can do without the checksum and only have ccid and cid? So contribution and contact are matched? Would that work?

Reply:
@magnolia61 I don't think so, cid needs a checksum to be authenticated, else contact id is not assigned while [checking permission](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Form.php#L2000-L2010) on loading of the form.

Reply:
@eileenmcnaughton Use-case mentioned by pradeep could be another reason to have a cid and checksum on pay now link for scenarios 1 & 4 other than having uniform link mentioned by @magnolia61 ?